### PR TITLE
Make cookie encoding conform better to RFC 6265

### PR DESCRIPTION
**Pull request by lukehutch**
Applies two changes to cookie encoding in strict mode:

1. ClientCookieEncoder will sort cookies in decreasing order of path length.
2. ServerCookieEncoder will only send one cookie of a given name (arbitrarily chosen as the last cookie set with the given name, ignoring path -- which cookie to send in the case of multiple cookies with the same name is not specified in the spec).

(Please double-check everything looks right, I admittedly have not tested this, as I don't have the bandwith right now to figure out how to build and incorporate a modified version of Netty into my project.)

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java [hunk 1]
```diff
@@ -142,16 +158,28 @@ public String encode(Cookie cookie) {
      * @return the corresponding bunch of Set-Cookie headers
      */
     public List<String> encode(Collection<? extends Cookie> cookies) {
-        if (checkNotNull(cookies, "cookies").isEmpty()) {
+        if (checkNotNull(cookies, "cookies").size() == 0) {
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java]: Why did you change this?

Comment (lukehutch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java]: Good catch, I initially removed the whole method, since it seemed to be subsumed by the version that took an Iterable (since Collection implements Iterable). Then I realized that there was merit in having a version that takes Collection, because the .size() method can be used to pre-allocate the map to the right size. So I put the method back again, but since it was short, I recreated it manually rather than going back and finding the original.
(In summary, this should still read ".isEmpty()")

Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java]: cool :)

**Comment by normanmaurer** [comment]
@lukehutch let me know once this is ready for review and once you have fixed all checkstyle-errors.

**Comment by slandelle** [comment]
Side comment: what Netty still lacks on the client side is a cookie jar. Maybe I'll someday find time to contribute one out of [Gatling's one](https://github.com/gatling/gatling/blob/master/gatling-http/src/main/scala/io/gatling/http/cookie/CookieJar.scala) (it's in Scala).

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 2]
```diff
@@ -91,12 +121,66 @@ public String encode(Cookie... cookies) {
         }
 
         StringBuilder buf = stringBuilder();
-        for (Cookie c : cookies) {
-            if (c == null) {
-                break;
+        if (strict) {
+            ArrayList<Cookie> cookiesList = new ArrayList<Cookie>(cookies.length);
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: List<Cookie> indeed ;)
https://docs.oracle.com/javase/6/docs/api/java/util/Collections.html#sort(java.util.List, java.util.Comparator)

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 3]
```diff
@@ -91,12 +121,66 @@ public String encode(Cookie... cookies) {
...
+     * @return a Rfc6265 style Cookie header value, null if no cookies are passed.
+     */
+    public String encode(Collection<? extends Cookie> cookies) {
+        if (checkNotNull(cookies, "cookies").size() == 0) {
+            return null;
+        }
+
+        StringBuilder buf = stringBuilder();
+        if (strict) {
+            ArrayList<Cookie> cookiesList = new ArrayList<Cookie>(cookies.size());
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: List

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 4]
```diff
@@ -114,13 +198,30 @@ public String encode(Iterable<? extends Cookie> cookies) {
         }
 
         StringBuilder buf = stringBuilder();
-        while (cookiesIt.hasNext()) {
-            Cookie c = cookiesIt.next();
-            if (c == null) {
-                break;
+        if (strict) {
+            ArrayList<Cookie> cookiesList = new ArrayList<Cookie>();
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: List

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java [hunk 5]
```diff
@@ -163,17 +191,32 @@ public String encode(Cookie cookie) {
...
+        if (strict) {
+            Map<String, Cookie> nameToCookie = new HashMap<>();
+            while (cookiesIt.hasNext()) {
+                Cookie c = cookiesIt.next();
+                if (c == null) {
+                    break;
+                }
+                nameToCookie.put(c.name(), encode(c));
+            }            
+            return new ArrayList<>(nameToCookie.values());
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java]: Mmm... The method returns a List there, which somehow implies that it keeps the Iterable input order. That's no longer the case with your change.

I wonder if returning a List is really useful, in which case your PR needs to be fixed, or if it doesn't, in which case we should change the method signature to return a Collection, and get rid of the extra allocation and directly return the Map values.

**Comment by slandelle** [comment]
@lukehutch Being curious: did you actually encountered a server that would break when the client doesn't order cookies this way?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java (hunk 2 above)
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: You should be directly using an array here and use Arrays.sort.
This way, you would save an allocation (Collections.sort copies into an internal array and delegate to Arrays.sort).

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 6]
```diff
@@ -91,12 +121,66 @@ public String encode(Cookie... cookies) {
...
+        if (strict) {
+            ArrayList<Cookie> cookiesList = new ArrayList<Cookie>(cookies.length);
+            for (Cookie c : cookies) {
+                if (c == null) {
+                    break;
+                }
+
+                cookiesList.add(c);
             }
+            Collections.sort(cookiesList, COOKIE_COMPARATOR);
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Arrays.sort

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java (hunk 3 above)
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: array

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 7]
```diff
@@ -91,12 +121,66 @@ public String encode(Cookie... cookies) {
...
+        if (strict) {
+            ArrayList<Cookie> cookiesList = new ArrayList<Cookie>(cookies.size());
+            for (Cookie c : cookies) {
+                if (c == null) {
+                    break;
+                }
+
+                cookiesList.add(c);
+            }
+            Collections.sort(cookiesList, COOKIE_COMPARATOR);
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Arrays.sort

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 8]
```diff
@@ -91,12 +127,64 @@ public String encode(Cookie... cookies) {
...
+            List<Cookie> cookiesList = new ArrayList<Cookie>(cookies.length);
+            for (Cookie c : cookies) {
+                if (c == null) {
+                    break;
+                }
+
+                cookiesList.add(c);
+            }
+            Collections.sort(cookiesList, COOKIE_COMPARATOR);
+            for (Cookie c : cookiesList) {
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: If we iterator over an array, we could save an Iterator here and use a good old for (int i = 0; ...)

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java (hunk 2 above)
Comment (lukehutch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Oh, you're right, thanks. I always assumed that you couldn't sort a List, since it doesn't have guaranteed O(1) access time for get(). Makes sense that it makes an interim copy to get around this issue.

However, I don't understand the purpose of the following lines -- how important is it to be able to terminate the cookie list by including a null cookie?

```java
    if (c == null) {
        break;
    }
```

The reason I ask is because to make the change to use an array for sorting, rather than a list, I need to allocate an array and then copy the objects in from the list (or collection or iterable). However, since the list can be terminated by a single null object, I would need to resize the array if a null object is encountered in the list. Can I safely remove the above lines from each of the methods? (The behavior of taking no further cookies after a null cookie is encountered is not documented in the method comments anyway.)

**Comment by slandelle** [comment]
@lukehutch It seems I was still commenting while you were updating your PR, so some of my comments got "hidden". Did you get them?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java (hunk 5 above)
Comment (lukehutch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java]: Yes, the order is preserved in a newer commit. Thanks. However, you are right that the order doesn't matter, since the returned values will each be written to separate HTTP headers, and the header order is not important.

Unit tests would have to be modified to be order-invariant, and it's possible that the unexpected order may cause user consternation, but see if the efficiency overhead is worth it in the newer commit (which preserves order).

**Comment by slandelle** [comment]
@lukehutch Arghhh, I just received your comments by email, but I can't find them on Github?! So I can't reply on the spot.

Regarding the `if (c == null) break;`, I realize that something that was inherited from the [previous implementation](https://github.com/netty/netty/blob/3.9/src/main/java/org/jboss/netty/handler/codec/http/cookie/ServerCookieEncoder.java#L135-L137). I don't get it either and I think it should be turned into just skipping it. Nice catch.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java (hunk 7 above)
Comment (lukehutch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: @slandelle: I didn't encounter a broken server, I'm building my own server on top of Netty, and was trying to get cookie behavior right according to the spec. I had my own cookie classes, but decided to throw them away and use raw Netty cookies, then I noticed that the Netty classes don't adhere to the spec.

The changes in the patch matter for APIs that take or return a single cookie by name, e.g. for cookie deletion (where maxAge is set to 0). This will give unexpected results if there are multiple cookies with the same name, in arbitrary order, each with different paths. You can only set (i.e. delete) one cookie at a time in the response according to the spec, and you have to pick which one to delete first if you are asked to delete a cookie based on name alone. Also it's nice to have an API that returns a single cookie with a requested name (this should be the cookie with the longest path), as opposed to returning a list of cookies, which the user has to deal with.

(I have encountered maddening problems in the past when my cookie code wasn't quite spec-compliant for other reasons, so I think it's important to get right.)

There may be other things that need fixing in the Netty code to make it spec-compliant, I don't know -- I have only been looking closely at this one issue of cookie duplication.

**Comment by lukehutch** [comment]
Yes, I unfolded all your hidden comments. I'll remove the `if (c == null) break;` code then, I don't think it makes sense to have it there. Will check in another version with array sorting soon.

**Comment by lukehutch** [comment]
PS @normanmaurer / @slandelle major apologies for the crummy quality of the initial PR -- this started out as what I thought would be a minor change, so I was just typing code into github in my browser. I have the code building and the tests running in Eclipse now -- please excuse the initial laziness!

**Comment by lukehutch** [comment]
@slandelle, please take another look.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 9]
```diff
@@ -80,6 +88,34 @@ public String encode(Cookie cookie) {
...
+            String path2 = c2.path();
+            // Cookies with unspecified path default to the path of the request. We don't
+            // know the request path here, but we assume that the length of an unspecified
+            // path is longer than any specified path (i.e. pathless cookies come first),
+            // because setting cookies with a path longer than the request path is of
+            // limited use.
+            int len1 = path1 == null ? -1 : path1.length();
+            int len2 = path2 == null ? -1 : path2.length();
+            int assumedLen1 = len1 == -1 ? len2 + 1 : len1;
+            int assumedLen2 = len2 == -1 ? len1 + 1 : len2;
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Why not:
```
int len1 = path1 == null ? Integer.MAX_VALUE : path1.length();
int len2 = path2 == null ? Integer.MAX_VALUE : path2.length();
```

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java [hunk 10]
```diff
@@ -15,9 +15,11 @@
  */
 package io.netty.handler.codec.http.cookie;
 
-import static io.netty.handler.codec.http.cookie.CookieUtil.*;
+import static io.netty.handler.codec.http.cookie.CookieUtil.add;
+import static io.netty.handler.codec.http.cookie.CookieUtil.addQuoted;
+import static io.netty.handler.codec.http.cookie.CookieUtil.stringBuilder;
+import static io.netty.handler.codec.http.cookie.CookieUtil.stripTrailingSeparator;
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java]: @normanmaurer What's the coding style rule regarding static imports wildcard usage here?

**Comment by slandelle** [comment]
@lukehutch You really must add tests demonstrating path ordering and cookie dedup.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java (hunk 9 above)
Comment (lukehutch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Good suggestion, done.

**Comment by lukehutch** [comment]
@slandelle I did add tests for both of these already -- they should be in the same PR as of the second commit. (In the client case, I modified a test so that it passes with the correct ordering; in the server case I added two new tests.)

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java (hunk 10 above)
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java]: I usually not use wildcard imports. But there is no rule.

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java [hunk 11]
```diff
@@ -37,7 +37,9 @@ public void testEncodingMultipleClientCookies() {
         cookie2.setSecure(false);
         Cookie cookie3 = new DefaultCookie("myCookie3", "myValue3");
         String encodedCookie = ClientCookieEncoder.STRICT.encode(cookie, cookie2, cookie3);
-        assertEquals(c1 + c2 + c3, encodedCookie);
+        // Cookies should be sorted into decreasing order of path length, as per RFC6295.
+        // Where no path is provided, we assume maximum path length (so cookie3 comes first).
```
Comment (slandelle) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java]: When

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 12]
```diff
@@ -91,12 +127,52 @@ public String encode(Cookie... cookies) {
...
+        }
 
-            encode(buf, c);
+        StringBuilder buf = stringBuilder();
+        if (strict) {
+            Cookie[] cookiesSorted = new Cookie[cookies.size()];
+            int i = 0;
+            for (Cookie c : cookies) {
+                cookiesSorted[i++] = c;
+            }
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Just use:

```java
Cookie[] cookiesSorted = cookies.toArray(new Cookies[cookies.size()]); 
```

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 13]
```diff
@@ -114,13 +190,27 @@ public String encode(Iterable<? extends Cookie> cookies) {
...
-            if (c == null) {
-                break;
+        if (strict) {
+            List<Cookie> cookiesList = new ArrayList<Cookie>();
+            while (cookiesIt.hasNext()) {
+                Cookie c = cookiesIt.next();
+                cookiesList.add(c);
             }
+            Cookie[] cookiesSorted = new Cookie[cookiesList.size()];
+            cookiesList.toArray(cookiesSorted);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Use the same style as mention above.

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java (hunk 11 above)
Comment (lukehutch) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java]: What do you mean by "When"?

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java [hunk 14]
```diff
@@ -37,7 +37,9 @@ public void testEncodingMultipleClientCookies() {
...
+        assertEquals(c3 + "; " + c2 + "; " + c1, encodedCookie);
```
Comment (slandelle) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java]: None of those cookies have a path, so why is the order reversed?

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java (hunk 11 above)
Comment (slandelle) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java]: I mean the comment should be "WHEN no path is provided", not "WHERE no path is provided"

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieEncoderTest.java [hunk 15]
```diff
@@ -60,4 +63,29 @@ public void testEncodingWithNoCookies() {
...
+        List<String> result = new ArrayList<String>();
+        result.add("cookie1=value1");
+        result.add("cookie2=value2");
+        result.add("cookie1=value3");
+        Cookie cookie1 = new DefaultCookie("cookie1", "value1");
+        Cookie cookie2 = new DefaultCookie("cookie2", "value2");
+        Cookie cookie3 = new DefaultCookie("cookie1", "value3");
+        List<String> encodedCookies = ServerCookieEncoder.LAX.encode(cookie1, cookie2, cookie3);
+        assertEquals(result, encodedCookies);
+    }
```
Comment (slandelle) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieEncoderTest.java]: Correct me if I'm wrong, but it looks to me there's no test with cookies with non null path that would demonstrate the issue.

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java (hunk 14 above)
Comment (lukehutch) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ClientCookieEncoderTest.java]: cookie1 and cookie2 do have a path -- see the `.setPath()` lines. cookie3 does not have a path. I made a judgment call in sorting path-less cookies before cookies with paths, because the sort order is supposed to be most specific to most general (longest path to shortest path, in general), and path-less cookies inherit the path of the current request.

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieEncoderTest.java (hunk 15 above)
Comment (lukehutch) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieEncoderTest.java]: This is for server cookie encoding, not client encoding. There are two different things going on in this PR:

1. For client encoding, paths are sorted in decreasing order of path length in strict mode.
2. For server encoding, cookies are not sorted in strict mode, but if there is more than one cookie with the same name, all but the last instance is dropped (i.e. the last-set cookie with a matching name is the only one that is encoded in the header). This is another judgment call I made, since the spec does not say which cookie should be chosen in the case where there are multiple cookies with the same name, it only says that only one of them should be sent. There may be a better heuristic involving paths or path lengths, but picking the last-set cookie of a given name is arguably as reasonable as anything else.

This test does test case 2. The first test I added is for strict mode, the second test is for lax mode.

Comment (slandelle) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieEncoderTest.java]: Sorry, lack of sleep :)

Comment (lukehutch) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieEncoderTest.java]: > Sorry, lack of sleep :)

Me too, I'm kind of embarrassed about my initial PR!

**Comment by lukehutch** [comment]
I don't understand the last failure -- What can cause a 403 Forbidden when applying a patch?

**Comment by normanmaurer** [comment]
@lukehutch I think it is not related but some issue with github. Can you please just squash all your commits into one with the proper commit message ?

**Comment by lukehutch** [comment]
@normanmaurer: Sorry, git n00b here. I tried to do this using the rebase
suggestion in http://stackoverflow.com/a/30227725/3950982 (see
commit e2962d376bc9902c5dcb5d5f3238d256b5630ae5), but I'm not sure if it
worked. Let me know if I need to do something different.

**Comment by normanmaurer** [comment]
@lukehutch usually you do `git rebase -i HEAD^^^^^^` (use as much ^ as commits you want to squash) and then just replace `pick` with `f` for these that you want to merge in the previous commit.

**Comment by slandelle** [comment]
git rebase -i HEAD~numberOfCommit
change pick into r for commits you want to rename and f for commits you want to fix up
then git push origin branchName -f

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 16]
```diff
@@ -91,12 +127,48 @@ public String encode(Cookie... cookies) {
         }
 
         StringBuilder buf = stringBuilder();
-        for (Cookie c : cookies) {
-            if (c == null) {
-                break;
+        if (strict) {
+            Cookie[] cookiesSorted = Arrays.copyOf(cookies, cookies.length);
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Add short path when there's only one cookie?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 17]
```diff
@@ -91,12 +127,48 @@ public String encode(Cookie... cookies) {
...
+     * @return a Rfc6265 style Cookie header value, null if no cookies are passed.
+     */
+    public String encode(Collection<? extends Cookie> cookies) {
+        if (checkNotNull(cookies, "cookies").isEmpty()) {
+            return null;
+        }
+
+        StringBuilder buf = stringBuilder();
+        if (strict) {
+            Cookie[] cookiesSorted = cookies.toArray(new Cookie[cookies.size()]);
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Add short path when there's only one cookie?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 18]
```diff
@@ -91,12 +127,48 @@ public String encode(Cookie... cookies) {
...
+            Arrays.sort(cookiesSorted, COOKIE_COMPARATOR);
+            for (Cookie c : cookiesSorted) {
+                encode(buf, c);
+            }
+        } else {
+            for (Cookie c : cookies) {
+                if (c == null) {
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: remove break block

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 19]
```diff
@@ -114,13 +186,26 @@ public String encode(Iterable<? extends Cookie> cookies) {
         }
 
         StringBuilder buf = stringBuilder();
-        while (cookiesIt.hasNext()) {
-            Cookie c = cookiesIt.next();
-            if (c == null) {
-                break;
+        if (strict) {
+            List<Cookie> cookiesList = new ArrayList<Cookie>();
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: short path when there's only one cookie

**Comment by slandelle** [comment]
@lukehutch Could you also fix the PR name, please? That's not RFC6295 but RFC6265 ;-)

**Comment by lukehutch** [comment]
@normanmaurer OK, I collapsed the PR down to one request, sorry for the commit spam.

@slandelle I added fast paths for single-element arrays/lists/Iterables, and fixed the PR name.

Please take another look. Thanks for all the good feedback.

**Comment by slandelle** [comment]
@lukehutch @normanmaurer that was fun :)

**Comment by normanmaurer** [comment]
@lukehutch sorry for be a PITA but two last things. Could you please alter the commit message to match our commit template:

http://netty.io/wiki/writing-a-commit-message.html

You can do this by using:
```
git commit --amend
```
Then do a force push again via:
```
git push -f origin branchName
```

After that please sign our ICLA:
http://netty.io/s/icla

After both things are done I will cherry-pick the change.  Thanks again for all your effort!

**Comment by lukehutch** [comment]
@slandelle I added fast paths for the case of one cookie to the server code too, PTAL. (The code is pretty straightforward, except in the case of Iterable, which is a little more complex.)

@normanmaurer I altered the commit message into the requested format, and signed the ICLA.

Thanks guys for your patience and feedback!

**Comment by lukehutch** [comment]
(Just made a couple of last minor changes -- improved a variable name and fixed a comment. I think I'm done with the patch now, unless you have other suggestions or requests.)

**Comment by lukehutch** [comment]
Also, I see that the milestone is set to 4.0.34.Final, but this patch is against HEAD, which I assume is 5.0. Do I need to also prep a patch for the 4.0 branch, or will you do that?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java [hunk 20]
```diff
@@ -91,12 +130,53 @@ public String encode(Cookie... cookies) {
...
+    public String encode(Collection<? extends Cookie> cookies) {
+        if (checkNotNull(cookies, "cookies").isEmpty()) {
+            return null;
+        }
+
+        StringBuilder buf = stringBuilder();
+        if (strict) {
+            if (cookies.size() == 1) {
+                for (Cookie c : cookies) {
+                    encode(buf, c);
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: encode(buf, cookies.iterator().next())

**Comment by slandelle** [comment]
@lukehutch Norman will cherry-pick on the different branches

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java (hunk 20 above)
Comment (lukehutch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java]: Thanks, nice trick.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (4154ea08f9293bdd6eca6bbb3c7d9c649c08c1c8), 4.1 (4978266d52a90252ae00b40894a4398292830d7f) and master (13af429e150ee62808541f2b1e8143c54f64d0ad)

**Comment by normanmaurer** [comment]
@lukehutch thanks for the PR... @slandelle thanks for review

**Comment by lukehutch** [comment]
Thank you @normanmaurer and @slandelle for all your feedback, and sorry for the bumpy ride, it was a good learning experience!

**Comment by slandelle** [comment]
@lukehutch @normanmaurer Thank you. That's how open source should work :)
